### PR TITLE
Handle new format for public asset resources

### PR DESCRIPTION
**Pull request by ClementPasteau**
The main goal is to allow easy modification of existing public assets, without impacting projects that may be using them.

The strategy picked is to add a unique id to the files, corresponding the sha of the file (thus, when it changes, the file changes without affecting the existing one)

This PR is to handle those new file formats.

### CODE REVIEW

File: newIDE/app/src/Utils/GDevelopServices/Asset.js [hunk 1]
```diff
@@ -328,3 +332,30 @@ export const listReceivedAssetPacks = async (
...
+const escapeStringForRegExp = string =>
+  string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
+const resourceFilenameRegex = new RegExp(
+  `(${escapeStringForRegExp(
+    GDevelopPublicAssetResourcesStorageBaseUrl
+  )}|${escapeStringForRegExp(
+    GDevelopPublicAssetResourcesStorageStagingBaseUrl
+  )})\\/public-resources\\/(.*)\\/([a-z0-9]{64})_(.*)(\\..*)`
+);
+export const extractFilenameAndExtensionFromPublicAssetResourceUrl = (
```
Comment (4ian) [inline newIDE/app/src/Utils/GDevelopServices/Asset.js:348]: Something that I wanted to do in my still not merged PR about cloud projects on desktop:
don't return "filename and extension", just return the filename with the extension (`myfile.png`).

The caller can extract the extension if they wish (but it's not always the case, at one point the code is concatenating them just after getting them!)

Comment (ClementPasteau) [reply newIDE/app/src/Utils/GDevelopServices/Asset.js:348]: true, I'll change that.

File: newIDE/app/src/ProjectsStorage/ResourceFetcher/index.js [hunk 2]
```diff
@@ -39,14 +57,47 @@ type UseResourceFetcherOutput = {|
...
-    options: FetchAllProjectResourcesOptionsWithoutProgress
+    options: CleanAndFetchAllProjectResourcesOptionsWithoutProgress
   ) => Promise<void>,
   /**
    * Render, if needed, the dialog that will show the progress of resources fetching.
    */
   renderResourceFetcherDialog: () => React.Node,
 |};
 
+export const cleanUpResourceNames = ({
```
Comment (ClementPasteau) [inline newIDE/app/src/ProjectsStorage/ResourceFetcher/index.js:68]: I've tried creating this function to call it in both fetchers.
The goal is to clean up the names of the resources (and remove the possible unwanted token & sha) before asking the fetcher to fetch it.
Wdyt of the strategy? @4ian

File: newIDE/app/src/ProjectsStorage/LocalFileStorageProvider/LocalFileResourceMover.js [hunk 3]
```diff
@@ -53,25 +53,45 @@ export const moveUrlResourcesToLocalFiles = async ({
       const resource = resourcesManager.getResource(resourceName);
 
       const url = resource.getFile();
-      const {
-        extension,
-        filenameWithoutExtension,
-      } = extractFilenameAndExtensionFromProductAuthorizedUrl(url);
+      let filenameWithExtension;
+      if (isProductAuthorizedResourceUrl(url)) {
```
Comment (4ian) [inline newIDE/app/src/ProjectsStorage/LocalFileStorageProvider/LocalFileResourceMover.js:57]: Very clear

File: newIDE/app/src/ProjectsStorage/ResourceFetcher/index.js [hunk 4]
```diff
@@ -39,14 +57,47 @@ type UseResourceFetcherOutput = {|
...
+        resourceFile
+      );
+    } else if (isPublicAssetResourceUrl(resourceFile)) {
+      // Resource is a public asset.
+      cleanedUpFilename = extractFilenameWithExtensionFromPublicAssetResourceUrl(
+        resourceFile
+      );
+    } else {
+      // Resource is a generic url.
+      cleanedUpFilename = path.basename(resourceFile);
```
Comment (4ian) [inline newIDE/app/src/ProjectsStorage/ResourceFetcher/index.js:95]: Could we just have this being factored into a function, and reused whenever need in the various resource fetchers?

**Review by 4ian — APPROVED** [review]

**Pull request merged**